packages/router/router.ts

```diff
@@ -3647,21 +3788,33 @@ function getMatchesToLoad(
   fetchRedirectIds: Set<string>,
   routesToUse: AgnosticDataRouteObject[],
   basename: string | undefined,
-  pendingActionData?: RouteData,
-  pendingError?: RouteData
+  pendingActionResult?: PendingActionResult
 ): [AgnosticDataRouteMatch[], RevalidatingFetcher[]] {
-  let actionResult = pendingError
-    ? Object.values(pendingError)[0]
-    : pendingActionData
-    ? Object.values(pendingActionData)[0]
+  let actionResult = pendingActionResult
+    ? isErrorResult(pendingActionResult[1])
+      ? pendingActionResult[1].error
+      : pendingActionResult[1].data
     : undefined;
-
   let currentUrl = history.createURL(state.location);
   let nextUrl = history.createURL(location);
 
   // Pick navigation matches that are net-new or qualify for revalidation
-  let boundaryId = pendingError ? Object.keys(pendingError)[0] : undefined;
-  let boundaryMatches = getLoaderMatchesUntilBoundary(matches, boundaryId);
+  let boundaryId =
+    pendingActionResult && isErrorResult(pendingActionResult[1])
+      ? pendingActionResult[0]
+      : undefined;
+  let boundaryMatches = boundaryId
+    ? getLoaderMatchesUntilBoundary(matches, boundaryId)
+    : matches;
+
+  // Don't revalidate loaders by default after action 4xx/5xx responses
+  // when the flag is enabled.  They can still opt-into revalidation via
+  // `shouldRevalidate` via `actionResult`
+  let shouldSkipRevalidation =
+    skipActionErrorRevalidation &&
+    pendingActionResult &&
+    typeof pendingActionResult[1].statusCode === "number" &&
+    pendingActionResult[1].statusCode >= 400;
```

Comment:
Don't revalidate if `unstable_skipActionErrorrRevalidation` is enabled and the action result is a 4xx/5xx response